File: leetcode/509.fibonacci-number.cpp
/*
 * @lc app=leetcode id=509 lang=cpp
 *
 * [509] Fibonacci Number
 */

// @lc code=start
#include <vector>
using namespace std;

class Solution
{
public:
    int fib(int n)
    {
        if (n <= 0)
            return n;

        vector<int> dp(n + 1, 0);

        dp[0] = 0;
        dp[1] = 1;
        for (int i = 2; i <= n; i++)
        {
            dp[i] = dp[i - 1] + dp[i - 2];
        }

        return dp[n];
    }
};
// @lc code=end

File: leetcode/509.fibonacci-number.py
#
# @lc app=leetcode id=509 lang=python3
#
# [509] Fibonacci Number
#

# @lc code=start
class Solution:
    def fib(self, n: int) -> int:
        dp = [-1] * n+1

        def fibSelf(n: int) -> int:
            if n <= 1:
                return n

            if dp[n] != -1:
                return dp[n]

            return fibSelf(n - 1) + fibSelf(n - 2)

        return fibSelf(n)
# @lc code=end

